File: 0049-group-anagrams/0049-group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> map = new HashMap<>();
        
        for (int i = 0; i < strs.length; i++) {
            String cur = orderedString(strs[i]);
            if (map.containsKey(cur)) {
                List<String> curList = map.get(cur);
                curList.add(strs[i]);
            } else {
                List<String> newList = new ArrayList<>();
                newList.add(strs[i]);
                map.put(cur, newList);
            }
        }

        List<List<String>> result = new ArrayList<>();
        for (String key : map.keySet()) {
            result.add(map.get(key));
        }

        return result;
    }

    public String orderedString(String s) {
        char[] sArr = s.toCharArray();
        Arrays.sort(sArr);
        return new String(sArr);
    }
}

File: 0049-group-anagrams/0049-group-anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> anagramMap;
        for (auto& str : strs) {
            string sortedStr = getSortedCopy(str);
            anagramMap[sortedStr].push_back(str);
        }
        
        vector<vector<string>> result;
        for (auto& p : anagramMap) {
            result.push_back(p.second);
        }
        
        return result;
    }

private:
    string getSortedCopy(string str) {
        sort(str.begin(), str.end());
        return str;
    }
};